File: src/main/java/mods/eln/item/BrushDescriptor.java
package mods.eln.item;

import mods.eln.generic.GenericItemUsingDamageDescriptor;
import mods.eln.misc.Utils;
import mods.eln.misc.UtilsClient;
import mods.eln.wiki.Data;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.client.IItemRenderer;
import org.lwjgl.opengl.GL11;

import java.util.List;

import static mods.eln.i18n.I18N.tr;

public class BrushDescriptor extends GenericItemUsingDamageDescriptor {

	private final ResourceLocation icon;
	private static ResourceLocation dryOverlay = new ResourceLocation("eln", "textures/items/brushdryoverlay.png");

	public BrushDescriptor(String name) {
		super( name);
		icon = new ResourceLocation("eln", "textures/items/" + name.toLowerCase().replace(" ", "") + ".png");
	}

	@Override
	public String getName(ItemStack stack) {
		int color = getColor(stack),life = getLife(stack);
		if(color == 15 && life == 0)
			return "Empty " + super.getName(stack);
		return super.getName(stack);
	}

	@Override
	public void setParent(Item item, int damage) {
		super.setParent(item, damage);
		Data.addWiring(newItemStack());
	}

	public int getColor(ItemStack stack) {
		return stack.getItemDamage() & 0xF;
	}

	public int getLife(ItemStack stack) {
		return stack.getTagCompound().getInteger("life");
	}

	public void setColor(ItemStack stack,int color) {
		stack.setItemDamage((stack.getItemDamage() & ~0xF) | color);
	}

	public void setLife(ItemStack stack,int life) {
		stack.getTagCompound().setInteger("life", life);
	}
	
	@Override
	public NBTTagCompound getDefaultNBT() {
		NBTTagCompound nbt = new NBTTagCompound();
		nbt.setInteger("life", 32);
		return nbt;
	}
	
	@Override
	public ItemStack newItemStack(int size) {
		return super.newItemStack(size);
	}

	@Override
	public void addInformation(ItemStack itemStack, EntityPlayer entityPlayer, List list, boolean par4) {
		super.addInformation(itemStack, entityPlayer, list, par4);
		
		if (itemStack != null)
			list.add(tr("Can paint %1$ blocks", itemStack.getTagCompound().getInteger("life")));
	}
	
	public boolean use(ItemStack stack, EntityPlayer entityPlayer) {
		int life = stack.getTagCompound().getInteger("life");
		if(life != 0) {
			--life;
			stack.getTagCompound().setInteger("life", life);
			return true;
		}
		else
			Utils.addChatMessage(entityPlayer, tr("Brush is dry"));
		return false;
	}

	@Override
	public boolean use2DIcon() {
		return true;
	}

	@Override
	public boolean handleRenderType(ItemStack item, IItemRenderer.ItemRenderType type) {
		return type == IItemRenderer.ItemRenderType.INVENTORY;
	}

	@Override
	public boolean shouldUseRenderHelper(IItemRenderer.ItemRenderType type, ItemStack item, IItemRenderer.ItemRendererHelper helper) {
		return type != IItemRenderer.ItemRenderType.INVENTORY;
	}

	@Override
	public void renderItem(IItemRenderer.ItemRenderType type, ItemStack item, Object... data) {
		if (type == IItemRenderer.ItemRenderType.INVENTORY) {
			UtilsClient.drawIcon(type, icon);
			GL11.glColor4f(1, 1, 1, 0.75f - 0.75f * getLife(item) / 32f);
			UtilsClient.drawIcon(type, dryOverlay);
			GL11.glColor3f(1, 1, 1);
		} else {
			super.renderItem(type, item, data);
		}
	}
}

File: src/main/java/mods/eln/item/BrushDescriptor.kt
package mods.eln.item

import mods.eln.generic.GenericItemUsingDamageDescriptor
import mods.eln.i18n.I18N.tr
import mods.eln.misc.Utils
import mods.eln.wiki.Data
import net.minecraft.client.Minecraft
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.nbt.NBTTagInt
import net.minecraft.util.ResourceLocation

class BrushDescriptor(name: String): GenericItemUsingDamageDescriptor(name) {

    private val icon = ResourceLocation("eln", "textures/items/" + name.toLowerCase().replace(" ", "") + ".png")

    override fun getName(stack: ItemStack): String {
        val creative = Minecraft.getMinecraft().player.capabilities.isCreativeMode
        val color = getColor(stack)
        val life = getLife(stack)
        return if (!creative && color == 15 && life == 0) "Empty " + super.getName(stack) else super.getName(stack)
    }

    override fun setParent(item: Item, damage: Int) {
        super.setParent(item, damage)
        Data.addWiring(newItemStack())
    }

    fun getColor(stack: ItemStack) = stack.itemDamage and 0xF

    private fun getLife(stack: ItemStack?): Int = stack?.tagCompound?.getInteger("life") ?: 32

    fun setLife(stack: ItemStack, life: Int) {
        stack.setTagInfo("life", NBTTagInt(life))
    }

    override fun getDefaultNBT(): NBTTagCompound? {
        val nbt = NBTTagCompound()
        nbt.setInteger("life", 32)
        return nbt
    }

    override fun addInformation(itemStack: ItemStack?, entityPlayer: EntityPlayer?, list: MutableList<Any?>, par4: Boolean) {
        super.addInformation(itemStack, entityPlayer, list, par4)

        if (itemStack != null) {
            val creative = Minecraft.getMinecraft().player.capabilities.isCreativeMode
            list.add(tr("Can paint %s blocks", if (creative) "infinite" else itemStack.tagCompound!!.getInteger("life")))
        }
    }

    fun use(stack: ItemStack, entityPlayer: EntityPlayer): Boolean {
        val creative = entityPlayer.capabilities.isCreativeMode
        var life = stack.tagCompound!!.getInteger("life")
        return if (creative || life != 0) {
            if (!creative) {
                --life
                stack.tagCompound!!.setInteger("life", life)
            }
            true
        } else {
            Utils.sendMessage(entityPlayer, tr("Brush is dry"))
            false
        }
    }

// TODO(1.10): Reimplement brush coloring
//    override fun renderItem(type: IItemRenderer.ItemRenderType, item: ItemStack, vararg data: Any) {
//        if (type == IItemRenderer.ItemRenderType.INVENTORY) {
//            val creative = Minecraft.getMinecraft().player.capabilities.isCreativeMode
//            UtilsClient.drawIcon(type, icon)
//            if (!creative) {
//                GL11.glColor4f(1f, 1f, 1f, 0.75f - 0.75f * getLife(item) / 32f)
//                UtilsClient.drawIcon(type, dryOverlay)
//                GL11.glColor3f(1f, 1f, 1f)
//            }
//        } else {
//            super.renderItem(type, item, *data)
//        }
//    }

    companion object {
        private val dryOverlay = ResourceLocation("eln", "textures/items/brushdryoverlay.png")
    }
}
